PyUnicode for Jython 3

jeff5 opened the pull request:
Well here's a monster. The PR provides `str` (`PyUnicode`) on the new core, and makes `PyString` disappear. `PyBytes` will be along eventually, using an actual `byte[]`. A few features are missing from this, like `str.__mod__` and anything that needs a real Python iterator (other PRs). But there's a heck of a lot here to chew on.

As before, I'll leave this hanging a week (at least) before pulling it into `main`.

I'm taking advantage of the way Jython 3 can treat different Java classes as first-class implementations of the same Python type. If you want proper Python behaviour for `str`, across the whole multilingual plane, only a custom class wrapping an array of `int` code points will cut it. But it would be convenient just to supply a `java.lang.String` to any method expecting a `str`. So `PyUnicode` and `String` are both `str`.

If you want to do something complicated with that Java `String` (as a `str`), you have to deal properly with the encoding of SMP characters as surrogate pairs. There are tests that we avoid, for example, `"\ud83e"+"\udd93" == "🦓"` in Python, even when both are `String` in Java.

A system of wrappers (sequence delegate classes) makes it possible to write the algorithms, for the many `str` methods, just once and have them apply to the `int[]` and `String` representations. Although it won't look much like it, the algorithms are mostly preserved from the Jython 2 implementations. It is possible behaviours are still Python 2-ish, and I know a few are missing. Test coverage of the methods is mostly poor: the Python test suite will catch mistakes.

For the purpose of traceability, I'm again making the transformation incrementally, and in this case that means a lot of intermediate stages where the build is broken. Normally I avoid merging broken states, but I will, so that the journey is visible in `main`. I'll only merge when the final state is good.

jeff5 committed:
[Build fails] Remove stop-gap PyUnicode

jeff5 committed:
[Build fails] Move Jython 2 PyUnicode to new core

We also bring in the hierarchy PyString and PyBaseString, although we
currently expect to delete them, once we have salvaged what is inherited
by PyUnicode. PyString is not a beginning for PyBytes. We're leaving
advanced formatting (str.format) out for now.

jeff5 committed:
[Build fails] Mark-up exposed methods

The crudely mechanical addition of comment markers around the Jython 2
annotations preserves information and makes the diffs more readable, so
others can follow when everything but the intent changes.

jeff5 committed:
[Build fails] Move PyString code into PyUnicode

There will be no instances of PyString, so this change is approximately
as follows. Everything is deleted from PyString that PyUnicode overrides
(in Java or Python) and does not call. Everything from PyString that
PyUnicode inherits or calls, or that we might still need, is moved to
PyUnicode. We create some stop-gap classes to satisfy references in the
code. (What we delete mistakenly is retrievable from git.)

This extensive code motion, in the interest of preserving prior art, is
yet not enough to make it all compile.

jeff5 committed:
[Build fails] Re-write str.find, split, replace and similar

We largely disentangle PyUnicode from Jython 2 PyString. We rename
methods in Jython 3 plain style although they are not tested at this
stage. The build fails on Jython 2 code, unconverted as we proceed by
stages.

We implement the methods that depend on searching for occurrences of a
substring, to replace or split. We add a test of these methods. (The
test passes in a version elsewhere without the Jython 2 code.)

jeff5 committed:
[Build fails] Many predicates and transformations in str

Another increment in the port of str, this change converts predicates
like isupper() and transformations like upper(). The new implementation
uses the system of delegates, iterators and builders already introduced,
but the logic is taken from Jython 2. The change includes a partial
implementation of str.join().

There is not much test coverage of these except for thorough tests of
str.join(). (The Python test suite will cover them once we can execute
code.)

jeff5 committed:
[Build fails] Implement the strip methods in str and simplify

We re-implement strip, lstrip and rstrip following the same approach as
Jython 2 but using the delegates and iterators as previously. We add a
fairly thorough test.

We are finally able to delete a lot of Jython 2 plumbing code. The
remaining code to convert is about formatting.

jeff5 committed:
[Build fails] Move Jython 2 format iterators to core

We move FieldNameIterator and MarkupIterator into core.stringlib from
the preserved Jython 2 sources, and the associated test, renamed as
FormatParsingTest.

They don't compile, but we want it to be clear where they came from, and
git loses track if we edit them in the same change.

jeff5 committed:
Port __format__ support to Jython 3

We convert the apparatus of formatting to the conventions of the new
core. str.__format__ is obviously affected, but also int.__format__ and
float.__format__. For the moment, str.__mod__ is removed, and when
re-instated will use the formatters in stringlib. With these changes, we
finally say goodbye to PyString and PyBaseString.

The rambling Jython 2 StringFormatTest is re-worked as two tests, each
with parameterised methods. The coverage is about the same.

Minor changes elsewhere are mostly to do with making visible objects
that we have kept package private.

jeff5 committed:
Make core:javadoc task run cleanly

Stewori commented:
(Disclaimer: without having had an actual look at it)

Wouldn't it be better to rather call it `PyString` and say that `PyUnicode` disappears? IIRC that is how CPython did it.

Did you consider support/integration with the Java interface `java.lang.CharSequence`?

jeff5 commented:
@Stewori (Happy New Year) `CharSequence` was very much on my mind because of your earlier remark. This is useful when we might pass the `str` implementation to a method (in Java obviously) that expects it. The last stages of this brought up some cases of that with `InternalFormat.Formatter` which works everywhere with `String` or `CharSequence`. (Overlook that it may be better ultimately if it were to use arrays of code points.) I dealt with that by coercing to String, which I can do fairly efficiently.

https://github.com/jeff5/jython/blob/5418dc4905458a045a2a1dec564f9ea3165f24e4/core/src/main/java/org/python/core/PyUnicode.java#L5287

I was put off by the question of what would `CharSequence.charAt(i)` return? Is `i` the UTF-16 index, or consistent with Python `self[i]`? I think the former, as I can't return a 32-bit code point, so to return it from a `PyUnicode` I have to encode the `int[]`. And the same may be asked about `CharSequence.substring()`, and even `CharSequence.length()`.

If it can't be done efficiently, it might be an attractive nuisance. I can see a possible mitigation for the inefficiency, by caching the `String` form if there are SMP elements. I thought I'd wait for more compelling use cases.

I considered `PyStr`, as the name, for consistency with `str `, and `PyInt `for `int`. Of course, there are no classes in C, but I've gone with `PyUnicode` and `PyLong` to match the prefixes you find in CPython on functions with a similar purpose.

https://github.com/python/cpython/blob/7c5b01b5101923fc38274c491bd55239ee9f0416/Objects/unicodeobject.c#L15166

Elsewhere you'll find `PySequence` and `PyNumber` classes holding static methods just so I can write [PyNumber.add(29, 13)](https://docs.python.org/3/c-api/number.html?highlight=pynumber_add#c.PyNumber_Add). Yes, that actually works!

After 5 weeks of quite hard work `PyUnicode` still has gaps. There are some other things that I want to prove possible in Jython 3, more than I want to perfect `str`, e.g. that we can interpret CPython byte code. The standard I'm aiming for right now is that `str` be good enough for that, and that it preserve the value of prior contributions.

jeff5 commented:
Although the substance of this is as much as I want to do in the PR, I have some refactoring in mind, and a strengthened test.

Stewori commented:
Happy new year indeed! Thank you, Jeff, for considering everything so carefully. I somehow thought CPython had abandoned ``unicodeobject.c`` and merged it into ``stringobject.c``, probably I just assumed that to match what happened with the datatypes ``unicode`` and ``str``. Still, a Jython implementation without a class ``PyString`` feels odd to me (for no reason).

> what would CharSequence.charAt(i) return

From Javadoc of `CharSequence.charAt`:

> Returns the char value at the specified index. An index ranges from zero to length() - 1. The first char value of the sequence is at index zero, the next at index one, and so on, as for array indexing.
>
> If the char value specified by the index is a surrogate, the surrogate value is returned.

Do I understand correctly that this implies it is the UTF-16 index? Otherwise it would not make sense to mention surrogates, right?  Consequently, `length()` returns the UTF-16 encoded length while `len(str)` would return the possibly different logical length of a string? That is not satisfying, `length` should be consistent with `len`. Maybe it should just throw an ``UnsupportedOperationException``, at least if surrogates are present :/
What about instead matching Python behavior and returning for supplementary characters the higher or lower (whatever you think makes more sense) surrogate? At least this would be efficient.
Okay, after re-reading parts of your introduction, given that PyUnicode and also String can serve as str, maybe the behavior of ``charAt`` and ``length`` should simply match the underlying implementation "dumbly". Obviously, we don't have control anyway if it is a String. The indexing should just be forwarded to the underlying array or buffer. Sure, this would delegate the actual issue to the user, but I never had in mind to introduce much magic or deeper logic with supporting `CharSequence`. It has always been just an interfacing idea, in terms of data typing, making available what is already there in a more compatible, standardized fashion. If that is not feasible for reasons I was not aware of, just drop the idea.

Triaging implementation priority is absolutely wise and I am glad you are planning accordingly.

jeff5 committed:
Factor out the int array builders from PyUnicode

This non-functional change anticipates re-use and reduces the size of
PyUnicode. There are some side-effects of the re-factoring, notably to
remove the specific takeUnicode in favour of one returning an array, to
introduce an indexed get(int) into PySequence.Of, and to increase
visibility of supportive API.

jeff5 committed:
Test that a Python str supports sequence operations

Within the limits of what is implemented, test sequence operations
through the Abstract API. Tests include size, index, slice,
concatenation, and repeat. We don't have str iteration yet.

We verify with SMP strings in both implementations where there seems a
risk of incorrect behaviour.

jeff5 commented:
I agree that charAt would have to return a UTF-16 unit and with:

> Consequently, `length()` returns the UTF-16 encoded length while `len(str)` would return the possibly different logical length of a string? That is not satisfying, `length` should be consistent with `len`. 

But consistent is not the same. `len("🦓")` is one character while `"🦓".length()` in Java is two code units. You can't execute that in Python, since `str` has no method `length`. There will be some weirdness between Python's view and Java's APIs, for example `StringBuilder("🦓").length() == 2`.

> Maybe it should just throw an `UnsupportedOperationException`, at least if surrogates are present :/ 

Since a user would have no real control over the representation, an exception would be very annoying as described there. In writing a much longer reply than this now is, I realised it is impossible to provide a `CharSequence` representation of certain Python `str` values, Consider:
```
>>> s = "\ud83d\udc0d"
>>> s == '🐍'
False
>>> len(s)
2
```
Java does not (I think) have any way to encode `s`. The obvious choice, maybe the only choice, would be as two surrogates which it would then treat as a pair representing  '🐍'. Maybe an exception is the only way out at this point. It will come up (at least) in `__tojava__`, when we're trying to pass a `str` to a `String` parameter but it has a _value_ Java can't represent, therefore the called method can't accept in that position. So I think it is a Python `ValueError` we throw (raise into Python).

Stewori commented:
> Java does not (I think) have any way to encode `s`

Why not? `s` can be encoded in UTF-16, right? Is Java's support for UTF-16 incomplete? I used to think that Java's `String` precisely consists of UTF-16 content.

Regarding the other aspects I agree that `PyUnicode` probably should not implement `CharSequence` since that interface is (contrary to its name) not an abstraction for char sequences but specifically for UTF-16 content.
Only thing left to consider would be whether maybe generally `CharSequence` (rather than just `String`) can be accepted as `str`.

jeff5 commented:
> > Java does not (I think) have any way to encode `s`
> 
> Why not? `s` can be encoded in UTF-16, right? Is Java's support for UTF-16 incomplete? I used to think that Java's `String` precisely consists of UTF-16 content.

Actually, no. (About the encoding.)
```
>>> a = array.array('H', '🐍'.encode('UTF-16LE'))
>>> s = "".join(map(chr, a))
>>> s
'\ud83d\udc0d'
>>> s.encode('UTF-16LE')
Traceback (most recent call last):
  File "<pyshell#379>", line 1, in <module>
    s.encode('UTF-16LE')
UnicodeEncodeError: 'utf-16-le' codec can't encode character '\ud83d' in position 0: surrogates not allowed
```
It isn't a valid character. The "obvious" encoding (each surrogate as the same value) encodes something else, namely the SMP code point. Java can store lone surrogates, although incorrectly.

Thanks to the discussion, I'm clear now that to pass a `str` to Java is, strictly-speaking, to encode it. It's just that sometimes we'll keep our `str` in its encoded form.

We should probably have a  constructor from `CharSequence` but we can't accept it as an implementation without the guarantee of immutability.

Stewori commented:
I still don't get it. If UTF-16 allows potentially two 16-bit characters to encode a single unicode character, why wasn't this done such that it actually can encode the whole unicode set? Is this a bug?
I read the (apparently outdated) article https://www.oracle.com/technical-resources/articles/javase/supplementary.html but it does not mention this issue in a single word. According to that article, every unicode character should be representable in UTF-16.
From [Wikipedia](https://en.wikipedia.org/wiki/UTF-16) (not outdated, right?):

> UTF-16 (16-bit Unicode Transformation Format) is a character encoding capable of encoding all 1,112,064 valid character code points of Unicode (in fact this number of code points is dictated by the design of UTF-16).

UTF-16LE means much the same as UTF-16 in terms of representation capability, just pinning the char order to little endian, right?

So what is the issue now?

* '🐍' is not a valid unicode character
* bug in Java (and Python UTF-16 codec)
* Wikipedia is incorrect

jeff5 commented:
> I still don't get it. ... From [Wikipedia](https://en.wikipedia.org/wiki/UTF-16) (not outdated, right?):
> 
> > UTF-16 (16-bit Unicode Transformation Format) is a character encoding capable of encoding all 1,112,064 valid character code points of Unicode (in fact this number of code points is dictated by the design of UTF-16).

> So what is the issue now?

'\ud83d' is [not a valid character](https://en.wikipedia.org/wiki/UTF-16#U+D800_to_U+DFFF). UTF-16 shouldn't encode it (YMMV) and it definitely can't distinguish `'\ud83d\udc0d'`from '🐍', although Python can.

It's an issue if Python throws one of those pairings at Java. Arguably they shouldn't arise.

Stewori commented:
Then '🐍' is not a valid unicode character, right? Is it misleading to talk of it as "Unicode" in the first place? It should be considered a Python bug that it does accept it (Edit: I see, the encoding function actually rejects it, so no bug there. Before that function call it is just unchecked data. Still not sure if I understand the problem correctly. Anyway, thanks for your patience in explaining.)

Stewori commented:
What happens if one just puts`'\ud83d'`, `'\udc0d'` in Java into a `char[]` and feeds that to `String(char[] value)`, not caring to encode it "actually" as UTF-16? That's what you mean by

> Java can store lone surrogates, although incorrectly.

, right?

If Java eats that and there is a feasible way to get that data from the `PyUnicode` data I (personally) would prefer that "incorrect" way over a `ValueError`. The incorrectness would originate on Python side after all and this attempt would just be best effort to carry it on, assuming that is what the user wants.

I still do not really understand why there are characters assigned to invalid codepoints. 1,112,064 valid character code points should be enough "legal" space, also for snakes (although it is disappointingly small compared to the 32 bit a character occupies).

jeff5 committed:
Fix typos in comments and some formatting

jeff5 committed:
Add test of our UTF-16 expectations

This addition follows review of PR #147, so we can check our assumptions
on Java handling of supplementary characters, and especially of lone
surrogates, in String, and when encoding to UTF-16. This is important to
choices about integration of String with Python str.

jeff5 commented:
'🐍' is a perfectly acceptable Unicode character U+1F40D that encodes to 0xd83d, 0xdc0d. It is correct to deliver it to as "\ud83d\udc0d" to a Java method with a  `String` (or `CharSequence`) parameter. Also, I can type "🐍" in my IDE (where the text is being stored as UTF-8), but the `String` the compiler produces has two UTF-16 units "\ud83d\udc0d". This is all fine.

Both Python and Java are able to store isolated surrogates. However, Python is able to store isolated surrogates even when they occur in the order that would encode a SMP character. Java also stores them separately, of course, because it stores UTF-16 units, but that combination represents the SMP character (and it should appear as one in Python when `str` is implemented by `String`).

I notice that encoding a lone surrogate (where Python raises an error) does not throw but delivers a UNICODE REPLACEMENT character (encoded in UTF-16). This may be preferable to `ValueError` in `__tojava__`, when I come to revisit that.

I'm adding a test (https://github.com/jython/jython/pull/147/commits/87ea2d543f1e3ffe2b420f9299dc6df9697428ba) that exercises these assumptions to save repeat research. It is good to have bottomed this out.

Stewori commented:
Sorry Jeff, I still don't understand the issue.

> '🐍' is a perfectly acceptable Unicode character U+1F40D

Then no lone surrogate should be required, no? (because lone surrogates are invalid UTF-16 while for every acceptable Unicode character there exists a proper UTF-16 representation) Why would s.encode('UTF-16LE') error then and not just produce the proper UTF-16 representation?

Is the error that somehow UTF-16BE content is interpreted as UTF-16LE? But that would be an erroneous setting in the first place.

jeff5 commented:
There isn't much of an issue, and none with `'🐍'`, when that's what you mean to represent.

It is a minor issue that Python can represent `'\ud83d\udc0d'` and `'🐍'` as distinct values of `str`, while Java cannot represent them distinctly as `String`s.

I am clear that a Java `"\ud83d\udc0d"` should be considered a Python `'🐍'`, and that a Python `'🐍'` should be converted to a Java `"\ud83d\udc0d"` whenever a `CharSequence` is required, e.g. in `__tojava__`.

I think converting a Python `'\ud83d\udc0d'` to a Java `String` should produce some kind of error, because this `str` represents a pair of surrogate code points that the programmer intends to handle as "lone". Even though they occur in the right order to fuse, they don't mean `'🐍'`, because a `str` is not a UTF-16 encoding result, but an array of code points.

I think this will only arise when calling a Java method from Python. The most Java-like solution is surely the default behaviour of the Java UTF-16 codec, which treats any surrogate code point as not mappable to UTF-16, and to encode the Unicode replacement character in its place. The most Pythonic is an exception. Converting to a Java `"\ud83d\udc0d"` seems a bad idea because it silently changes the meaning.
